port/osx/omrosdump.c

```diff
@@ -0,0 +1,388 @@
+		file_off += sizeof(x86_exception_state_t);
+		mh64->sizeofcmds += threads[i].cmdsize;
+	}
+	mh64->ncmds += thread_count;
+
+	kr = list_segment_commands(task_port, &segments, &segment_count);
+	if (kr != KERN_SUCCESS) {
+		mach_error("error getting thread command data:\n", kr);
+		goto done;
+	}
+	seg_file_off = file_off + segment_count * sizeof(struct segment_command_64);
+	for (int i = 0; i < segment_count; i++) {
+		vm_offset_t data_read;
+		mach_msg_type_number_t data_size;
+
+		segments[i].fileoff = seg_file_off;
+		pwrite(corefile_fd, &segments[i], sizeof(struct segment_command_64), file_off);
+
+		kr = mach_vm_read(task_port, segments[i].vmaddr, segments[i].vmsize, &data_read, &data_size);
+		ssize_t written;
+		written = pwrite(corefile_fd, (void *)data_read, data_size, seg_file_off);
+		mach_vm_deallocate(task_port, data_read, data_size);
+
+		file_off += sizeof(struct segment_command_64);
+		seg_file_off += segments[i].vmsize;
+	}
+	mh64->ncmds += segment_count;
+	mh64->sizeofcmds += segment_count * sizeof(struct segment_command_64);
+
+	/* write mach header after all command number and size are known */
+	pwrite(corefile_fd, mh64, sizeof(struct mach_header_64), 0);
+
+done:
+	if (corefile_fd > 0) {
+		close(corefile_fd);
+		if (kr != KERN_SUCCESS) {
+			remove(corefile_name);
+		}
+	}
+	return kr;
+}
+
+static int list_thread_commands(mach_port_t task_port, struct thread_command_full_64 **thread_commands, natural_t *thread_count) {
+	kern_return_t kr = KERN_SUCCESS;
+	thread_act_array_t thread_info;
+	struct thread_command_full_64 *threads = NULL;
+
+	kr = task_threads(task_port, &thread_info, thread_count);
+
+	if (kr != KERN_SUCCESS) {
+		mach_error("task_thread failed with: ", kr);
+		return kr;
+	}
+
+	threads = calloc(*thread_count, sizeof(struct thread_command_full_64));
+	if (NULL == threads) {
+		kr = KERN_NO_SPACE;
+		goto done;
+	}
+
+	for (int i = 0; i < *thread_count; i++) {
+		uint32_t state_int_count;
+		threads[i].cmd = LC_THREAD;
+		threads[i].cmdsize = 8;
+		state_int_count = x86_THREAD_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_THREAD_STATE, (thread_state_t)&threads[i].thread_state, &state_int_count);
+		if (kr != KERN_SUCCESS) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * 4;
+		state_int_count = x86_FLOAT_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_FLOAT_STATE, (thread_state_t)&threads[i].float_state, &state_int_count);
+		if (kr != KERN_SUCCESS) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * 4;
+		state_int_count = x86_EXCEPTION_STATE_COUNT;
+		kr = thread_get_state(thread_info[i], x86_EXCEPTION_STATE, (thread_state_t)&threads[i].exceptions, &state_int_count);
+		if (kr != KERN_SUCCESS) {
+			goto done;
+		}
+		threads[i].cmdsize += state_int_count * 4;
+	}
+	if (KERN_SUCCESS == kr) {
+		*thread_commands = threads;
+	}
+
+done:
+	for (int i = 0; i < *thread_count; i++) {
+		mach_port_deallocate(mach_task_self(), thread_info[i]);
```

Comment:
If a failure occurred above, `thread_info[i]` may not have been initialized.

Reply:
`thread_info` and all its elements are initialized in L221. The function exits if `task_threads` fails there.

Reply:
I understand that `thread_info` is initialized: I'm referring to individual elements of that array. My point was that one of the system calls in the loop beginning on line 234 might fail part way through the list. `thread_info[i]` is not meaningful for later values of `i`. Perhaps you need to track the number of successful iterations of that loop.

Reply:
I understand what you mean now, however the loop at L234 only reads from the `thread_info` array, and doesn't set anything. `task_threads` returns an array of thread ports in the `thread_info`, and the size of the array in `thread_count`, so every individual element of the array is set in this step.